// @progress/kendo-react-pdf/treelist/provideSaveTreeListPDF.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as a from "react";
import * as E from "react-dom";
import * as h from "react-dom/client";
import { exportElement as y } from "./export-element.mjs";
function P(c) {
  return (r, i = {}, s, l, n) => x(
    c,
    r,
    i,
    s,
    l,
    n
  );
}
function x(c, r, i = {}, s, l, n) {
  let t, o, e;
  f();
  function f() {
    t = document.createElement("div"), t.setAttribute("style", "position:absolute; left: -5000px; top: 0px;"), o = document.createElement("div"), t.appendChild(o), document.body.appendChild(t), Number(a.version.slice(0, 2)) >= 18 ? (h.createRoot(o).render(m()), setTimeout(() => {
      d();
    }, 0)) : E.render(m(), o, d);
  }
  function d() {
    c(v(), i, b);
  }
  function b() {
    document.body.removeChild(t), document.body.removeChild(e), t = e = void 0, s && s();
  }
  function v() {
    e = document.createElement("div"), e.className = "k-treelist-pdf-export-element";
    const p = y(o);
    return e.appendChild(p), document.body.appendChild(e), e;
  }
  function m() {
    const p = i.allPages && l ? { data: l, take: Number.MAX_VALUE, skip: 0 } : {}, C = { style: Object.assign({}, r.props.style, { width: "1000px" }) }, u = Object.assign({}, p, C);
    return n && n.length > 0 ? a.cloneElement(r, Object.assign({}, u, { columns: n })) : a.cloneElement(r, u);
  }
}
export {
  P as provideSaveTreeListPDF
};
