// @progress/kendo-react-pdf/treelist/TreeListPDFExport.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as i from "react";
import e from "prop-types";
import { getPageMargin as a } from "../getPageMargin.mjs";
import { savePDF as p } from "../savePDF.mjs";
import { provideSaveTreeListPDF as g } from "./provideSaveTreeListPDF.mjs";
const t = class t extends i.Component {
  constructor(r) {
    super(r), this.saveTreeListPDF = g(this.getSavePDF());
  }
  /**
   * @hidden
   */
  render() {
    return null;
  }
  /* eslint-disable max-len */
  /**
   * Saves the content of the TreeList as a PDF file.
   *
   * @param data - The data that will be exported to the PDF. The data can be different from the currently displayed data in the TreeList.
   * @param columns - The columns that will be exported to the PDF. The columns collection can be different from the currently displayed in the TreeList.
   * @param callback - The callback that will be executed after the PDF is saved.
   */
  // tslint:enable:max-line-length
  save(r, n, o) {
    this.saveTreeListPDF(
      this.getTreeList(),
      Object.assign({}, this.props, { margin: a(this.props) }),
      o,
      r,
      n || []
    );
  }
  getSavePDF() {
    return p;
  }
  getTreeList() {
    const r = this.props.children;
    if (r && r.props.data && r.props.columns)
      return this.props.children;
  }
};
t.propTypes = {
  author: e.string,
  avoidLinks: e.oneOfType([e.bool, e.string]),
  forcePageBreak: e.string,
  keepTogether: e.string,
  creator: e.string,
  date: e.instanceOf(Date),
  imageResolution: e.number,
  fileName: e.string,
  forceProxy: e.bool,
  keywords: e.string,
  landscape: e.bool,
  margin: e.oneOfType([e.string, e.number, e.shape({
    left: e.oneOfType([e.number, e.string]),
    top: e.oneOfType([e.number, e.string]),
    right: e.oneOfType([e.number, e.string]),
    bottom: e.oneOfType([e.number, e.string])
  })]),
  pageTemplate: e.any,
  paperSize: e.any,
  repeatHeaders: e.bool,
  scale: e.number,
  proxyData: e.any,
  proxyURL: e.string,
  proxyTarget: e.string,
  producer: e.string,
  subject: e.string,
  title: e.string,
  allPages: e.bool
};
let s = t;
export {
  s as TreeListPDFExport
};
